### PR TITLE
[feature] proxy: proxy could identify if the connection is from internal or external network

### PR DESCRIPTION
**Pull request by volgariver6**
internal or external network

## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #11761 

## What this PR does / why we need it:
We need to identify the client connection comes from internal network
or external network. Add a new config in proxy to define the internal
network CIDR. If the source IP is in it, then it is internal, otherwise, it
is external.

### CODE REVIEW

**Review by zhangxu19830126 — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

File: pkg/frontend/mysql_protocol.go [hunk 1]
```diff
@@ -2706,43 +2706,29 @@ func (mp *MysqlProtocolImpl) MakeEOFPayload(warnings, status uint16) []byte {
...
+				logError(mp.ses, mp.GetDebugString(), "failed to get extra info",
+					zap.Error(err))
+			}
+		}
 	} else {
-		mp.GetSession().requestLabel = label.Labels
-		logDebugf(mp.GetDebugString(), "got requested CN labels: %v", *label)
+		mp.SetSalt(extraInfo.Salt)
+		mp.GetSession().requestLabel = extraInfo.Label.Labels
+		mp.GetSession().internalConn = extraInfo.InternalConn
```
Comment (xzxiong) [inline pkg/frontend/mysql_protocol.go:2731]: 考虑到cn 可以在没有 proxy的情况下运行，建议改为: 0, 1, 2, ...。默认值对应unset
```suggestion
if extraInfo.InternalConn {
    mp.GetSession().connType = 1; // 0: unset; 1: private networks; 2: others
}
```

**Review by daviszhen — APPROVED** [review]

File: pkg/frontend/mysql_protocol.go (hunk 1 above)
Comment (volgariver6) [reply pkg/frontend/mysql_protocol.go]: done

File: pkg/proxy/util.go [hunk 2]
```diff
@@ -193,3 +194,17 @@ func rawHash(t any) string {
 	hash := md5.Sum(sortBytes)
 	return hex.EncodeToString(hash[:])
 }
+
+// containIP returns if the CIDR list contains the IP address.
+func containIP(cidrList []string, ip net.IP) bool {
+	for _, cidr := range cidrList {
+		_, ipNet, err := net.ParseCIDR(cidr)
```
Comment (xzxiong) [inline pkg/proxy/util.go:201]: 建议 benchmark 看下该函数的耗时，我理解网络连接对时延的要求挺高的。
优化建议：proxy只需要执行一轮 `net.ParseCIDR`，新连接只需要检查`ipNet.Contains`即可。

Comment (volgariver6) [reply pkg/proxy/util.go]: done

**Review by xzxiong — APPROVED** [review]
LGTM

**Pull request merged**